Page for reorgs and sequencing errors (#1277)
ethereum-optimism/docs: pages/stack/interop/reorg.mdx

The most frequent solution to mitigate the double-spend problem is to wait for Ethereum finality; however, that solution results in high latency cross-chain communication and a poor user experience.

<details>

<summary>What is double-spending?</summary>

```mermaid

flowchart LR 
    subgraph init ["Initiating transaction (source chain)"]
        burn(tokens burned)
        burn-->send(send)
    end
    subgraph exec ["Executing transaction (destination chain)"]
        send==initiating message==>receive(receive)
        receive-->mint(tokens minted)
    end
```

In a normal asset transfer tokens are burned on the source chain first, then a message is sent to the destination chain.
When that message is received, the tokens are minted on the destination chain, where the user can now use those tokens.

```mermaid

flowchart LR 
    subgraph init ["Not really the source chain"]
        err((error))
    end
    subgraph exec ["Executing transaction (destination chain)"]
        err==initiating message==>receive(receive)
        receive-->mint(tokens minted)
    end
```

> A double-spend problem occurs when the destination chain receives a valid initiating message, but due to an issue on the source chain, such as a reorg, that initiating transaction is no longer valid. 

Reviewer: ```suggestion
A double-spend problem occurs when the destination chain receives a valid initiating message, but due to issues on the source chain, such as a reorg, that initiating transaction is no longer valid. 
```

   Suggesting plural form as there could be multiple types of issues that cause this. @qbzzt

### L1 reorg

In this case, an L1 block is replaced, and the new block either contains the same blob with chain data that was contained in the old block, or it doesn't.
If the new block contains the same blob, the change is meaningless from the L2 perspective.
If the new block does not contain the same blob, then the sequencer on the chain will know it is missing and reposts it.

So L1 reorgs are basically invisible to L2.

### Equivocation

Sequencers inform the rest of the Superchain about a new block in two ways:

- The gossip protocol, which is typically used as soon as the block is created.
  The problem is that the gossip protocol does not create a commitment.
- Posting to L1, which typically happens a few minutes after the block is created.
  The reason is cost, it is a lot cheaper if compression and L1 posting are done in large batches, rather than for each individual block.

Equivocation happens when a sequencer publishes a block using the gossip protocol that is different from the one that eventually gets written to L1.

> In this case, the block that is written to L1 (let's call it A'<sub>101</sub>) is the valid one, and that causes every dependant block to be recalculated.

Reviewer: ```suggestion
In this case, the block that is written to L1 (let's call it A'<sub>101</sub>) is the valid one, and that causes every dependent block to be recalculated.
```

      B300-->B301b["B'<sub>301</sub>"]-->B302b["B'<sub>302</sub>"]-->B303b["B'<sub>303</sub>"]
    end
    subgraph C ["Chain C"]
      C200["C<sub>200</sub>"]-->C201["C<sub>201</sub>"]-->C202["C<sub>202</sub>"]-->C203["C<sub>203</sub>"]
      C202-->C203b["C'<sub>203</sub>"]
    end
    A101-.->B302
    A101-.->B301
    A101b-.->B302b
    A101b-.->B301b    
    B300-.->A100
    B302-.->C203
    B302b-.->C203b    
    C200-.->B301    
    class C200 finalized
    class A100,B300,C201,C202 safe
    class A101b,A102b,A103b,B301b,B302b,B303b,C203b safe
    class A101,A102,A103,B301,B302,B303,C203 unsafe
```

Note that a block can only depend directly on a block in a separate chain if the block includes an executing message that uses an initiating message from that block.
So the change from A<sub>101</sub> to A'<sub>101</sub> cannot invalidate any existing safe blocks.

### Invalid block

If a block is invalid, even if it is posted on L1, the canonical chain replaces it with a block that only includes the deposit transactions, those transactions posted to L1.

<details>

<summary>What makes a block invalid?</summary>

There are several potential reasons:

> - The block posted to L1 includes incorret information, for example because it relied on a node on a different blockchain for interop and that node reported incorrect information.

Reviewer: ```suggestion
- The block posted to L1 includes incorrect information, for example because it relied on a node on a different blockchain for interop and that node reported incorrect information.
```

---
title: Interop reorg awareness
lang: en-US

> description: How the OP Stack enables low-latency interop and avoids the double-spend problem.

Reviewer 2: How Superchain interop enables....

<details>

<summary>What is double-spending?</summary>

```mermaid

flowchart LR 
    subgraph init ["Initiating transaction (source chain)"]
        burn(tokens burned)
        burn-->send(send)
    end
    subgraph exec ["Executing transaction (destination chain)"]
        send==initiating message==>receive(receive)
        receive-->mint(tokens minted)
    end
```

In a normal asset transfer tokens are burned on the source chain first, then a message is sent to the destination chain.
When that message is received, the tokens are minted on the destination chain, where the user can now use those tokens.

```mermaid

flowchart LR 
    subgraph init ["Not really the source chain"]
        err((error))
    end
    subgraph exec ["Executing transaction (destination chain)"]
        err==initiating message==>receive(receive)
        receive-->mint(tokens minted)
    end
```

A double-spend problem occurs when the destination chain receives a valid initiating message, but due to an issue on the source chain, such as a reorg, that initiating transaction is no longer valid. 
When that happens, the tokens are still on the source chain, but they also on the destination chain.

</details>

> In Ethereum, incorrect state is usually a transient phenomenon.

Reviewer 2: What does "usually a transient phenomenon." mean?

> On L2, a sequencer can share one block through gossip and then post a different block to L1 (equivocation).

Reviewer 2: gossip protocol? is there a link to a spec or article that explains what this is?

Reviewer 2: To be honest isnt this paragraph just restating the first paragraph? consider nuking if so

Reviewer 2: basically we're trying to comm existing cross-chain solutions MUST wait for l1 finality to mitigate double spend. Superchain interop can mitigate double spend AND be low latency

If a sequencer equivocates, the dependent cross chain messages are reorged out once the block is posted by the sequencer to L1.
On L1, new block can be replaced through a reorg, but after sufficient time has passed the block is finalized, and is nearly impossible to remove from the blockchain.

The most frequent solution to mitigate the double-spend problem is to wait for L1 finality; however, that solution results in high latency and poor user experience.

## Block safety levels

> To mitigate the double-spend problem while delivering a low-latency cross-chain experience, OP Stack interop uses [block safety levels](./explainer#block-safety-levels).

Reviewer 2: Superchain interop uses...
